sql/src/test/java/io/crate/operation/scalar/AbstractScalarFunctionsTest.java

```diff
@@ -28,28 +28,101 @@
 import io.crate.types.DataType;
 import io.crate.types.DataTypes;
+import org.apache.lucene.util.BytesRef;
+import org.hamcrest.Matcher;
+import org.hamcrest.Matchers;
 import org.junit.Before;
 
 import java.util.Arrays;
 import java.util.List;
 
+import static org.hamcrest.core.Is.is;
+
 public abstract class AbstractScalarFunctionsTest extends CrateUnitTest {
     protected SqlExpressions sqlExpressions;
     protected Functions functions;
 
+
+    /**
+     * Assert that the functionExpression normalizes to the expectedSymbol
+     *
+     * If the result of normalize is a Literal and all arguments were Literals evaluate is also called and
+     * compared to the result of normalize - the resulting value of normalize must match evaluate.
+     */
+    @SuppressWarnings("unchecked")
+    public void assertNormalize(String functionExpression, Matcher<Symbol> expectedSymbol) {
+        Function function = (Function )sqlExpressions.asSymbol(functionExpression);
+        FunctionImplementation impl = functions.get(function.info().ident());
+
+        assertThat(impl, Matchers.notNullValue());
+
+        Symbol normalized = impl.normalizeSymbol(function);
+        assertThat(normalized, expectedSymbol);
+
+        if (normalized instanceof Input && allArgsAreInputs(function.arguments())) {
+            Input[] inputs = new Input[function.arguments().size()];
+            for (int i = 0; i < inputs.length; i++) {
+                inputs[i] = ((Input) function.arguments().get(i));
+            }
+            assertThat(((Scalar) impl).evaluate(inputs), is(((Input) normalized).value()));
+        }
+    }
+
+    /**
+     * asserts that the given functionExpression evaluates to the expectedValue.
+     * If the functionExpression contains references the inputs will be used in the order the references appear.
+     *
+     * E.g.
+     * <code>
+     *     assertEvaluate("foo(name, age)", "expectedValue", inputForName, inputForAge)
+     * </code>
+     * or
+     * <code>
+     *     assertEvaluate("foo('literalName', age)", "expectedValue", inputForAge)
+     * </code>
+     */
+    public void assertEvaluate(String functionExpression, Object expectedValue, Input ... inputs) {
+        Function function = (Function) sqlExpressions.asSymbol(functionExpression);
+        Scalar scalar = (Scalar) functions.get(function.info().ident());
+
+        Input[] arguments = new Input[function.arguments().size()];
+        int idx = 0;
+        for (int i = 0; i < function.arguments().size(); i++) {
+            Symbol arg = function.arguments().get(i);
+            if (arg instanceof Input) {
+                arguments[i] = ((Input) arg);
+            } else {
+                arguments[i] = inputs[idx];
+                idx++;
+            }
+        }
+        if (expectedValue instanceof String) {
+            expectedValue = new BytesRef((String) expectedValue);
+        }
+        assertThat(scalar.evaluate(arguments), is(expectedValue));
```

Comment:
Not necessarily, but I've added another assertion to make sure that the output of the compiled evaluate is the same as the un-compiled